src/ngraph_assign_clusters.cc

```diff
@@ -83,23 +91,203 @@ namespace {
+  if (!DeadnessAnalysis::IsTruePredString(src_predicate) &&
+      !DeadnessAnalysis::IsTruePredString(dst_predicate) &&
+      src_predicate != dst_predicate) {
+    is_deadness_ok = false;
+    return Status::OK();
+  }
+
+  // Case src X , dst True // invalid scenario
+  // If src has Non-True Predicate and dst has True Predicate, it implies that
+  // the dst node is control flow
+  if (!DeadnessAnalysis::IsTruePredString(src_predicate) &&
+      DeadnessAnalysis::IsTruePredString(dst_predicate)) {
+    return errors::Internal("Attempting to cluster control-flow node ",
+                            dst->name(), "[", dst->type_string(), "]");
+  }
+
+  // Case src True, dst Y
+  // If this edge is contracted, all the outputs of the merged cluster will
+  // have the predicate Y. Hence contraction is possible only when, all
+  // outputs of the src cluster (other than the current edge) have the
+  // predicate Y
+  if (DeadnessAnalysis::IsTruePredString(src_predicate)) {
+    auto src_cluster_out_edges = cluster_map[src]->outgoing_edges;
+    bool found_same_out_preds = true;
+    std::string pred_check = dst_predicate;
+
+    for (const Edge* src_cluster_edge : src_cluster_out_edges) {
+      if (src_cluster_edge == edge) {
+        continue;
+      }
+      Node* src_cluster_dst = src_cluster_edge->dst();
+      std::string src_cluster_dst_pred =
+          cluster_map[src_cluster_dst]->predicate_string;
+      if (pred_check != src_cluster_dst_pred) {
+        found_same_out_preds = false;
+        break;
+      }
+    }
+    // Cannot contract this edge
+    if (!found_same_out_preds) {
+      is_deadness_ok = false;
+      return Status::OK();
+    }
+  }
+
+  // Case src X, dst Y, X==Y
+  // Case src True, dst Y(Y can be True), all other output edges from src
+  // cluster have pred Y
+  // Ok to contract
+  is_deadness_ok = true;
+  return Status::OK();
 }
 
+// Some sanity checks for Node's cluster assignment wrt Deadness
+Status CheckNodeClusterAssignmentWRTDeadness(
+    Node* node, std::map<Node*, string>& nodes_predicate_map,
+    string& cluster_pred_string) {
+  if (nodes_predicate_map.find(node) == nodes_predicate_map.end()) {
+    return errors::Internal("Node ", node->name(), " [", node->type_string(),
+                            "]", " not found in predicate map");
+  }
+  std::string node_pred_string = nodes_predicate_map[node];
+
+  if (DeadnessAnalysis::IsControlFlowPredString(node_pred_string)) {
+    return errors::Internal(
+        "Node ", node->name(), " [", node->type_string(), "]",
+        " should not be clustered as it is a control flow op");
+  }
+
+  if (!DeadnessAnalysis::IsTruePredString(node_pred_string) &&
+      node_pred_string != cluster_pred_string) {
+    return errors::Internal(
+        "Node ", node->name(), " [", node->type_string(), "]", " Predicate : ",
+        node_pred_string,
+        "should not be clustered in cluster with pred_String ",
+        cluster_pred_string);
+  }
```

Comment:
In the current implementation, the bottom node can be True. For .e.g this graph merges into 1 cluster with pred P1. This is also one of the unit test case.
```
T --->P1
|
|
v
T----> P1 
|
|
v
P1
```
So, the outgoing edges can be either True or the Non-True Cluster Predicate P1. I will modify the check like this.

Reply:
I see, in your graph the bottom node is outside the cluster. So the test you suggested takes care of all scenarios. This is a valid Assert when the cluster pred of Node is Non-True (P1), which is evident in the notation, but not in the pseudocode. 
Reiterating for our benefit to avoid any misinterpretation or missing corner cases.
```
//consider a node named node
if (pred_node is True){
    P1 = predicate_of_cluster_node_is_in
    **if(P1 is not True){**
    for n in output_connection_nodes{
            if n is in same cluster as node{
                  //doesnt matter, continue.
            }
            else {
                node n must be in a cluster that has predicate == P1
            }
    }
}
}
```
Some more graphs 
Graph 1
```
T --> T
|   \
|    T 
v
P1
```
In this case there will be cluster of T nodes and an outgoing edge to P1.


Graph 2

T - - >* T
|   \
|    T 
v
P1

In this case there will be a cluster of True Nodes, 1 outgoing edge to True , and another to P1.